core/lib/snapshots_applier/src/lib.rs

```diff
@@ -0,0 +1,356 @@
+                )
+            }
+            _ => SnapshotsApplierError::Fatal(err.into()),
+        }
+    }
+}
+
+#[async_trait]
+pub trait SnapshotsApplierMainNodeClient: fmt::Debug + Send + Sync {
+    async fn fetch_l2_block(&self, number: MiniblockNumber) -> Result<Option<SyncBlock>, RpcError>;
+
+    async fn fetch_newest_snapshot(&self) -> Result<Option<SnapshotHeader>, RpcError>;
+}
+impl<'a> SnapshotsApplier<'a> {
+    pub async fn load_snapshot(
+        connection_pool: &'a ConnectionPool,
+        main_node_client: Box<dyn SnapshotsApplierMainNodeClient>,
+        blob_store: Box<dyn ObjectStore>,
+    ) -> anyhow::Result<(), SnapshotsApplierError> {
+        let latency =
+            METRICS.initial_stage_duration[&InitialStage::FetchMetadataFromMainNode].start();
+
+        let mut storage = connection_pool
+            .access_storage_tagged("snapshots_applier")
+            .await?;
+        let mut storage = storage.start_transaction().await?;
+
+        let mut applied_snapshot_status = storage
+            .snapshot_recovery_dal()
+            .get_applied_snapshot_status()
+            .await?;
+
+        if !storage.blocks_dal().is_genesis_needed().await? && applied_snapshot_status.is_none() {
+            return Err(SnapshotsApplierError::Canceled(
+                "This node has already been initialized without a snapshot".to_string(),
+            ));
+        }
+
+        if let Some(applied_snapshot_status) = applied_snapshot_status.as_ref() {
+            if !applied_snapshot_status
+                .storage_logs_chunks_processed
+                .contains(&false)
+            {
+                return Err(SnapshotsApplierError::Canceled(
+                    "This node has already been initialized from a snapshot".to_string(),
+                ));
+            }
+        } else {
+            applied_snapshot_status =
+                Some(SnapshotsApplier::create_fresh_recovery_status(main_node_client).await?)
+        }
+
+        let mut recovery = Self {
+            connection_pool,
+            blob_store,
+            applied_snapshot_status: applied_snapshot_status.unwrap(),
+        };
+
+        let latency = latency.observe();
+        tracing::info!("Initialized snapshots creator in {latency:?}");
+
+        METRICS.storage_logs_chunks_count.set(
+            recovery
+                .applied_snapshot_status
+                .storage_logs_chunks_processed
+                .len(),
+        );
+
+        METRICS.storage_logs_chunks_left_to_process.set(
+            recovery
+                .applied_snapshot_status
+                .storage_logs_chunks_processed
+                .iter()
+                .filter(|x| !(**x))
+                .count(),
+        );
+
+        recovery.recover_factory_deps(&mut storage).await?;
+
+        storage
+            .snapshot_recovery_dal()
+            .insert_initial_recovery_status(&recovery.applied_snapshot_status)
```

Comment:
This will error if the status is already present in Postgres (i.e., if it was loaded from it above).

Reply:
nice catch!